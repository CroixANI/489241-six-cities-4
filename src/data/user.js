class User {
  constructor(id, name, imageUrl, isPro) {
    this.id = id;
    this.name = name;
    this.imageUrl = imageUrl;
    this.isPro = isPro;
  }
}

const createUser = (hostData) =>
  new User(hostData.id, hostData.name, hostData.avatar_url, hostData.is_pro);

export {createUser};

export default User;
