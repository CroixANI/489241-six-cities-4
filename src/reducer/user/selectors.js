import NameSpace from '../name-space';

const NAME_SPACE = NameSpace.USER;

export const getAuthorizationStatus = (state) =>
  state[NAME_SPACE].authorizationStatus;

export const getIsAuthChecked = (state) =>
  state[NAME_SPACE].isAuthChecked;

export const getCurrentUserEmail = (state) => {
  const currentUser = state[NAME_SPACE].currentUser;
  if (currentUser) {
    return currentUser.email;
  }

  return ``;
};
