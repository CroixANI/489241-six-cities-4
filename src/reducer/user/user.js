import {createUser} from '../../data/user';

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  currentUser: null,
};

const ActionType = {
  REQUIRED_AUTHORIZATION: `REQUIRED_AUTHORIZATION`,
  SET_CURRENT_USER: `SET_CURRENT_USER`,
};

const ActionCreator = {
  requireAuthorization: (status) => {
    return {
      type: ActionType.REQUIRED_AUTHORIZATION,
      payload: status,
    };
  },
  setCurrentUser: (currentUser) => {
    return {
      type: ActionType.SET_CURRENT_USER,
      payload: currentUser,
    };
  },
};

const OperationCreator = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then(() => {
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      })
      .catch((err) => {
        throw err;
      });
  },
  login: (authData) => (dispatch, getState, api) => {
    return api.post(`/login`, {
      email: authData.login,
      password: authData.password,
    }).then((response) => {
      dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      dispatch(ActionCreator.setCurrentUser(createUser(response.data)));
    });
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.REQUIRED_AUTHORIZATION:
      return Object.assign({}, state, {
        authorizationStatus: action.payload
      });
    case ActionType.SET_CURRENT_USER:
      return Object.assign({}, state, {
        currentUser: action.payload
      });
  }

  return state;
};

export {reducer, ActionType, ActionCreator, OperationCreator, AuthorizationStatus};
