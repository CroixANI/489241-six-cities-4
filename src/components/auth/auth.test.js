import React from 'react';
import renderer from 'react-test-renderer';

import Auth from './auth.jsx';
import {AuthorizationStatus} from '../../reducer/user/user';

describe(`Render Auth component`, () => {
  it(`Render Auth component with NO_AUTH authorization status`, () => {
    const tree = renderer
      .create(<Auth authorizationStatus={AuthorizationStatus.NO_AUTH} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Render Auth component with AUTH authorization status`, () => {
    const tree = renderer
      .create(<Auth authorizationStatus={AuthorizationStatus.AUTH} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
