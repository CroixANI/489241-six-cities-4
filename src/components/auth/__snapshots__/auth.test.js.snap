// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Auth component Render Auth component with AUTH authorization status 1`] = `
<a
  className="header__nav-link header__nav-link--profile"
  href="/favorites"
  onClick={[Function]}
>
  <div
    className="header__avatar-wrapper user__avatar-wrapper"
  />
  <span
    className="header__user-name user__name"
  >
    Oliver.conner@gmail.com
  </span>
</a>
`;

exports[`Render Auth component Render Auth component with NO_AUTH authorization status 1`] = `
<a
  className="header__nav-link header__nav-link--profile"
  href="/login"
  onClick={[Function]}
>
  <div
    className="header__avatar-wrapper user__avatar-wrapper"
  />
  <span
    className="header__login"
  >
    Sign in
  </span>
</a>
`;
